File: src/RookNative/Handlers/ViewportHandler.cpp
// ViewportHandler.cpp
//
// POST /viewport — Capture the active viewport to a PNG file.
//
// Approach: Use _-ViewCaptureToFile via RunScript.
// This is simpler and more reliable than CRhinoDib in C++ since
// the command handles all display pipeline setup and file writing.
// The C# version uses view.CaptureToBitmap() (.NET only).
// Always returns a file path (no base64 — the MCP agent reads files directly).

#include "stdafx.h"
#include "Handlers/ViewportHandler.h"
#include "Handlers/GrasshopperProxyHandler.h"
#include "Infrastructure/JsonHelpers.h"
#include "Models/DocumentHelpers.h"
#include "Threading/MainThreadDispatcher.h"
#include "RookServer.h"

#include <filesystem>
#include <chrono>
#include <sstream>
#include <iomanip>

namespace fs = std::filesystem;

namespace Rook {
namespace Handlers {

// ─── Helpers ────────────────────────────────────────────────────────

// Generate a unique filename for the viewport capture.
static std::string GenerateFilename()
{
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()) % 1000;

    std::ostringstream oss;
    oss << "viewport_";

    struct tm tm_buf = {};
    if (localtime_s(&tm_buf, &time) == 0)
    {
        oss << std::put_time(&tm_buf, "%Y%m%d_%H%M%S");
    }
    else
    {
        oss << time;
    }

    oss << "_" << std::setfill('0') << std::setw(3) << ms.count() << ".png";
    return oss.str();
}

// Map standard view name to a Rhino _SetView command.
// Returns empty string if not a standard view name.
static std::wstring ViewNameToCommand(const std::string& viewName)
{
    if (IEquals(viewName, "top"))         return L"_-SetView _World _Top";
    if (IEquals(viewName, "bottom"))      return L"_-SetView _World _Bottom";
    if (IEquals(viewName, "front"))       return L"_-SetView _World _Front";
    if (IEquals(viewName, "back"))        return L"_-SetView _World _Back";
    if (IEquals(viewName, "left"))        return L"_-SetView _World _Left";
    if (IEquals(viewName, "right"))       return L"_-SetView _World _Right";
    if (IEquals(viewName, "perspective")) return L"_-SetView _World _Perspective";
    return {};
}

// ─── POST /viewport ─────────────────────────────────────────────────

void HandleViewport(const httplib::Request& req, httplib::Response& res)
{
    auto [docSn, body] = ParseBodyAndDocSn(req);

    // captureBackend discriminator: absent or "legacy" runs the in-process
    // native _-ViewCaptureToFile path; "tier3" dispatches to the managed
    // companion bridge (SDK-backed view.CaptureToBitmap with display-mode
    // resolution and optional raytraced convergence). Any other string is a
    // hard 400 — silent legacy fallback on a misspelling would be a false
    // success that consumers cannot diagnose.
    const std::string captureBackend = body.value("captureBackend", std::string());
    if (!captureBackend.empty()
        && captureBackend != "legacy"
        && captureBackend != "tier3")
    {
        CRookServer::SendError(res,
            "Unknown captureBackend '" + captureBackend +
            "'. Expected 'legacy' or 'tier3' (or omit the field).");
        res.status = 400;
        return;
    }

    if (captureBackend == "tier3")
    {
        // Forward the full request body to the managed Tier 3 handler. The
        // managed side reads documentSerialNumber from the same body and
        // pins the document via DocumentContext.WithDocument, so we do not
        // need to rewrite the payload — just pass it through.
        std::string responseJson;
        int statusCode = 0;
        std::string invokeError;
        const auto invokeResult = InvokeViewportCaptureTier3WithBody(
            req.body.empty() ? body.dump() : req.body,
            responseJson,
            statusCode,
            invokeError);

        switch (invokeResult)
        {
        case ManagedCreateInvokeResult::Ok:
            res.status = statusCode == 0 ? 200 : statusCode;
            res.set_content(responseJson, "application/json");
            res.set_header("X-Rook-Viewport-Backend", "tier3");
            return;
        case ManagedCreateInvokeResult::Unavailable:
            CRookServer::SendError(res,
                "Tier 3 viewport capture requires the Rook companion plugin. "
                "Ensure Rook.rhp is loaded in Rhino, then retry.");
            res.status = 503;
            res.set_header("X-Rook-Viewport-Backend", "tier3-unavailable");
            return;
        case ManagedCreateInvokeResult::Failed:
        default:
            CRookServer::SendError(res,
                "Tier 3 viewport capture failed: " + invokeError);
            res.status = 500;
            res.set_header("X-Rook-Viewport-Backend", "tier3-failed");
            return;
        }
    }

    // Parse parameters (all optional)
    int width = body.value("width", 800);
    int height = body.value("height", 600);
    std::string viewName = body.value("view", "");
    std::string displayMode = body.value("displayMode", "");
    bool zoomExtents = body.value("zoomExtents", false);
    bool transparentBackground = body.value("transparentBackground", false);
    int scale = body.value("scale", 1);
    bool drawGrid = body.value("drawGrid", false);
    bool drawWorldAxes = body.value("drawWorldAxes", false);
    bool drawCPlaneAxes = body.value("drawCPlaneAxes", false);

    // Clamp scale
    scale = (std::max)(1, (std::min)(scale, 10));

    // Clamp dimensions
    width = (std::max)(100, (std::min)(width, 4000));
    height = (std::max)(100, (std::min)(height, 4000));

    // Prepare output path
    fs::path tempDir = fs::temp_directory_path() / "rook" / "viewports";
    std::string filename = GenerateFilename();
    fs::path filePath = tempDir / filename;
    std::string filePathStr = filePath.string();

    auto future = CMainThreadDispatcher::Instance().Dispatch(
        [docSn, width, height, viewName, displayMode, zoomExtents,
         transparentBackground, scale, drawGrid, drawWorldAxes, drawCPlaneAxes,
         filePathStr]() -> nlohmann::json
    {
        CRhinoDoc* pDoc = ResolveDoc(docSn);
        unsigned int docRuntimeSn = pDoc->RuntimeSerialNumber();

        // Create output directory
        fs::create_directories(fs::path(filePathStr).parent_path());

        CRhinoView* view = pDoc->ActiveView();
        if (!view)
            throw std::runtime_error("No active view available");

        // Save current viewport state for restoration after capture.
        // RunScript commands permanently modify the viewport, so we must
        // restore the full ON_Viewport snapshot after capture.
        CRhinoViewport& vp = view->ActiveViewport();
        ON_Viewport savedVP = vp.VP();  // snapshot entire camera state
        bool viewChanged = false;
        bool displayModeChanged = false;

        std::string actualViewName = WideToUtf8(vp.Name());
        std::string actualDisplayMode = "current";

        // Apply view settings

        // 1. Set view if requested — standard view names or named views
        if (!viewName.empty())
        {
            std::wstring viewCmd = ViewNameToCommand(viewName);
            if (!viewCmd.empty())
            {
                // Standard view (Top, Front, etc.) via RunScript
                RhinoApp().RunScript(docRuntimeSn, viewCmd.c_str(), 0);
                actualViewName = viewName;
                viewChanged = true;
            }
            else
            {
                // Try as a named view — restore via SDK (Tier 1, no RunScript)
                ON_wString wName = Utf8ToWide(viewName);
                int namedViewCount = pDoc->Properties().NamedViewCount();
                int viewIdx = -1;
                for (int i = 0; i < namedViewCount; ++i)
                {
                    const ON_3dmView* pView = pDoc->Properties().NamedView(i);
                    if (pView && pView->m_name.CompareNoCase(wName) == 0)
                    {
                        viewIdx = i;
                        break;
                    }
                }
                if (viewIdx >= 0)
                {
                    const ON_3dmView* pNamedView = pDoc->Properties().NamedView(viewIdx);
                    if (pNamedView)
                    {
                        vp.SetVP(pNamedView->m_vp, true);
                        view->Redraw();
                        actualViewName = viewName;
                        viewChanged = true;
                    }
                }
                // If not found as named view either, silently use current view
            }
        }

        // 2. Set display mode if requested
        if (!displayMode.empty())
        {
            ON_wString wDisplayMode = Utf8ToWide(displayMode);
            std::wstring cmd = L"_-SetDisplayMode " +
                std::wstring(static_cast<const wchar_t*>(wDisplayMode));
            RhinoApp().RunScript(docRuntimeSn, cmd.c_str(), 0);
            actualDisplayMode = displayMode;
            displayModeChanged = true;
        }

        // 3. Zoom extents if requested
        if (zoomExtents)
        {
            RhinoApp().RunScript(docRuntimeSn, L"_-Zoom _All _Extents", 0);
            viewChanged = true;
        }

        // 4. Capture viewport to file using ViewCaptureToFile
        {
            ON_wString wFilePath = Utf8ToWide(filePathStr);
            std::wstring captureCmd = L"_-ViewCaptureToFile \"" +
                std::wstring(static_cast<const wchar_t*>(wFilePath)) +
                L"\" _Width=" + std::to_wstring(width) +
                L" _Height=" + std::to_wstring(height) +
                L" _Scale=" + std::to_wstring(scale) +
                L" _DrawGrid=" + (drawGrid ? L"Yes" : L"No") +
                L" _DrawWorldAxes=" + (drawWorldAxes ? L"Yes" : L"No") +
                L" _DrawCPlaneAxes=" + (drawCPlaneAxes ? L"Yes" : L"No") +
                L" _TransparentBackground=" + (transparentBackground ? L"Yes" : L"No") +
                L" _Enter";

            RhinoApp().RunScript(docRuntimeSn, captureCmd.c_str(), 0);
        }

        // 5. Restore viewport state if we changed it
        if (viewChanged)
        {
            CRhinoViewport& vpRestore = view->ActiveViewport();
            vpRestore.SetVP(savedVP, true);  // restore full camera/projection state
            view->Redraw();
        }

        // Verify the file was created
        if (!fs::exists(filePathStr))
            throw std::runtime_error("Failed to capture viewport");

        nlohmann::json data;
        data["format"] = "png";
        data["width"] = width;
        data["height"] = height;
        data["viewName"] = actualViewName;
        data["displayMode"] = actualDisplayMode;
        data["savedToFile"] = true;
        data["filePath"] = filePathStr;
        data["message"] = "Image saved to file. Use the Read tool to view the image.";
        return data;
    });

    try
    {
        auto data = future.get();
        CRookServer::SendSuccess(res, data);
    }
    catch (const std::exception& ex)
    {
        CRookServer::SendError(res, ex.what());
    }
}

} // namespace Handlers
} // namespace Rook

File: src/Rook/Handlers/ViewportHandler.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;
using Rhino;
using Rhino.Display;

namespace Rook.Handlers
{
    /// <summary>
    /// Tier 3 viewport capture — SDK-backed (<c>view.CaptureToBitmap</c>) with
    /// display-mode resolution and optional raytraced convergence. Invoked via
    /// the native bridge callback <c>viewport_capture_tier3</c> (ABI v13).
    ///
    /// The legacy Tier 2 path (<c>_-ViewCaptureToFile</c>) stays in native
    /// <c>/viewport</c>; this handler runs only when the caller sends
    /// <c>captureBackend: "tier3"</c>.
    ///
    /// Parity with legacy: full viewport state (camera + projection +
    /// display mode) is snapshotted before mutation and restored in
    /// <c>finally</c> — matches the native
    /// <c>ON_Viewport savedVP = vp.VP()</c> /
    /// <c>vp.SetVP(savedVP, true)</c> contract.
    ///
    /// Document pinning: the bridge trampoline wraps this call in
    /// <see cref="DocumentContext.WithDocument(uint?, Action)"/>, so
    /// <see cref="DocumentContext.GetDocument"/> returns the caller's
    /// target doc (multi-doc safe).
    ///
    /// Timeout contract: <c>raytracedTimeoutMs</c> is clamped to
    /// <see cref="MaxRaytracedTimeoutMs"/> (20 s) to leave headroom under
    /// the bridge callback's 30 s ceiling in
    /// <c>NativeGhBridgeRegistrar.ExecuteApiResponseCallback</c>.
    ///
    /// Scope: Tier 3 intentionally supports a narrower option set than the
    /// legacy <c>_-ViewCaptureToFile</c> path. The following legacy fields
    /// are rejected with HTTP 400 when present (rather than silently
    /// ignored): <c>scale</c>, <c>transparentBackground</c>,
    /// <c>drawGrid</c>, <c>drawWorldAxes</c>, <c>drawCPlaneAxes</c>.
    /// Silent degradation would produce false-success responses where the
    /// caller's flags do not affect the captured image. A future PR can
    /// map these onto <c>Rhino.Display.ViewCaptureSettings</c> after live
    /// verification of property-name semantics.
    /// </summary>
    public class ViewportHandler
    {
        internal const int DefaultRaytracedTimeoutMs = 10000;

        // Public cap stays comfortably under the bridge's 30 s waitHandle in
        // NativeGhBridgeRegistrar.ExecuteApiResponseCallback. Capture + PNG
        // write + JSON serialization needs the remaining ~10 s margin.
        internal const int MaxRaytracedTimeoutMs = 20000;

        internal const int RaytracedCheckIntervalMs = 250;
        internal const int RaytracedMinSamples = 10;
        internal const int RaytracedStableThreshold = 4;

        internal const int DefaultWidth = 800;
        internal const int DefaultHeight = 600;
        internal const int MinDimension = 100;
        internal const int MaxDimension = 4000;
        private static readonly string[] RaytracedModeNames =
            { "raytraced", "Raytraced", "cycles", "Cycles" };

        public ApiResponse CaptureTier3(string? body)
        {
            Tier3Request request;
            try
            {
                request = ParseRequest(body);
            }
            catch (ArgumentException ex)
            {
                return new ApiResponse { Success = false, Data = ex.Message };
            }

            var doc = DocumentContext.GetDocument();
            if (doc == null)
            {
                return new ApiResponse { Success = false, Data = "No active document." };
            }

            var view = doc.Views.ActiveView;
            if (view == null)
            {
                return new ApiResponse { Success = false, Data = "No active view available." };
            }

            // Snapshot the full viewport projection (camera + target + lens +
            // projection type) before any mutation, plus the display mode.
            // Matches the native handler's ON_Viewport savedVP = vp.VP()
            // save/restore contract at ViewportHandler.cpp:119 / :208.
            var savedProjection = new Rhino.DocObjects.ViewportInfo(view.ActiveViewport);
            var savedDisplayMode = view.ActiveViewport.DisplayMode;

            string resolvedViewName = view.ActiveViewport.Name;
            string resolvedDisplayMode =
                view.ActiveViewport.DisplayMode?.EnglishName ?? "current";
            int? raytracedSamples = null;

            try
            {
                // 1. View-token precedence MUST match legacy
                //    ViewportHandler.cpp (ViewNameToCommand first, then
                //    named-view fallback). Otherwise a document with a
                //    named view called "Top" captures different cameras
                //    depending on captureBackend — a silent regression.
                if (!string.IsNullOrEmpty(request.ViewName))
                {
                    if (TrySetStandardView(view, request.ViewName))
                    {
                        resolvedViewName = request.ViewName!;
                    }
                    else if (TryRestoreNamedView(doc, view, request.ViewName))
                    {
                        resolvedViewName = request.ViewName!;
                    }
                    // Unknown view name: silently keep the current view
                    // (matches legacy behavior).
                }

                // 2. Display mode: resolve via robust fallback chain and
                //    record what actually ran (not the caller token).
                bool runningRaytraced = false;
                if (!string.IsNullOrEmpty(request.DisplayMode)
                    && !string.Equals(request.DisplayMode, "current",
                                      StringComparison.OrdinalIgnoreCase))
                {
                    var targetMode = FindDisplayMode(request.DisplayMode!);
                    if (targetMode != null && targetMode.Id != savedDisplayMode?.Id)
                    {
                        view.ActiveViewport.DisplayMode = targetMode;
                        view.Redraw();
                        resolvedDisplayMode = targetMode.EnglishName;
                        runningRaytraced = IsRaytracedMode(targetMode);
                    }
                    else if (targetMode != null)
                    {
                        resolvedDisplayMode = targetMode.EnglishName;
                        runningRaytraced = IsRaytracedMode(targetMode);
                    }
                    // If targetMode == null, fall through to current mode
                    // silently — resolvedDisplayMode already reflects that.
                }
                else
                {
                    runningRaytraced = IsRaytracedMode(savedDisplayMode);
                }

                // 3. Zoom extents if requested.
                if (request.ZoomExtents)
                {
                    view.ActiveViewport.ZoomExtents();
                    view.Redraw();
                }

                // 4. Wait for raytraced convergence when both flag and mode
                //    agree. Silently ignored when either condition fails.
                if (request.RaytracedConverge && runningRaytraced)
                {
                    raytracedSamples = WaitForRaytracedRender(
                        view, request.RaytracedTimeoutMs);
                }

                // 5. Capture. Size = (width, height) clamped. SA_Banana's
                //    maxEdge helper is kept as internal scaling logic only
                //    and is NOT exposed on the public request contract.
                var captureSize = new Size(request.Width, request.Height);
                using var bitmap = view.CaptureToBitmap(captureSize);
                if (bitmap == null)
                {
                    return new ApiResponse
                    {
                        Success = false,
                        Data = "view.CaptureToBitmap returned null.",
                    };
                }

                // 6. Write to temp file, matching GrasshopperHandler.CaptureCanvasImage
                //    precedent (file-path transport, not base64).
                var tempDir = Path.Combine(
                    Path.GetTempPath(), "rook", "viewports");
                Directory.CreateDirectory(tempDir);
                var filename = "viewport_" +
                    DateTime.Now.ToString("yyyyMMdd_HHmmss_fff") + ".png";
                var filePath = Path.Combine(tempDir, filename);
                bitmap.Save(filePath, ImageFormat.Png);

                var data = new Dictionary<string, object?>
                {
                    ["format"] = "png",
                    ["width"] = bitmap.Width,
                    ["height"] = bitmap.Height,
                    ["viewName"] = resolvedViewName,
                    ["displayMode"] = resolvedDisplayMode,
                    ["savedToFile"] = true,
                    ["filePath"] = filePath,
                    ["message"] =
                        "Image saved to file. Use the Read tool to view the image.",
                    ["captureBackend"] = "tier3",
                };
                if (raytracedSamples.HasValue)
                {
                    data["raytracedSamples"] = raytracedSamples.Value;
                }

                return new ApiResponse { Success = true, Data = data };
            }
            catch (Exception ex)
            {
                return new ApiResponse
                {
                    Success = false,
                    Data = $"Tier 3 capture failed: {ex.Message}",
                };
            }
            finally
            {
                // Full viewport state restore — projection first (covers
                // camera, target, lens, view type), then display mode.
                try
                {
                    view.ActiveViewport.SetViewProjection(savedProjection, true);
                }
                catch (Exception ex)
                {
                    RhinoApp.WriteLine(
                        $"Rook: Tier 3 viewport projection restore failed: {ex.Message}");
                }
                try
                {
                    if (savedDisplayMode != null
                        && view.ActiveViewport.DisplayMode?.Id != savedDisplayMode.Id)
                    {
                        view.ActiveViewport.DisplayMode = savedDisplayMode;
                    }
                }
                catch (Exception ex)
                {
                    RhinoApp.WriteLine(
                        $"Rook: Tier 3 display mode restore failed: {ex.Message}");
                }
                try { view.Redraw(); } catch { /* swallow */ }
            }
        }

        // ─── Pure helpers (unit-testable without RhinoCommon) ──────────

        internal readonly struct Tier3Request
        {
            public int Width { get; init; }
            public int Height { get; init; }
            public string? ViewName { get; init; }
            public string? DisplayMode { get; init; }
            public bool ZoomExtents { get; init; }
            public bool RaytracedConverge { get; init; }
            public int RaytracedTimeoutMs { get; init; }
        }

        // Legacy /viewport accepts these. Tier 3 does not map them onto
        // view.CaptureToBitmap yet; accepting-and-ignoring would produce
        // false-success responses where the caller's flags do not affect
        // the captured image, so we hard-reject on presence.
        internal static readonly string[] UnsupportedLegacyFields =
        {
            "scale",
            "transparentBackground",
            "drawGrid",
            "drawWorldAxes",
            "drawCPlaneAxes",
        };

        internal static Tier3Request ParseRequest(string? body)
        {
            Dictionary<string, JsonElement>? args = null;
            if (!string.IsNullOrEmpty(body))
            {
                try
                {
                    args = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(body);
                }
                catch (JsonException ex)
                {
                    throw new ArgumentException($"Invalid JSON body: {ex.Message}");
                }
            }

            var backend = GetStringArg(args, "captureBackend");
            if (!string.IsNullOrEmpty(backend) && backend != "tier3")
            {
                // Legacy-path requests never reach this handler — native
                // branches on captureBackend before invoking the bridge.
                // A non-tier3 value here means caller bypassed /viewport,
                // which is still a protocol error.
                throw new ArgumentException(
                    $"ViewportHandler.CaptureTier3 requires captureBackend='tier3'. Got '{backend}'.");
            }

            if (args != null)
            {
                var rejected = new List<string>();
                foreach (var field in UnsupportedLegacyFields)
                {
                    if (args.ContainsKey(field))
                    {
                        rejected.Add(field);
                    }
                }
                if (rejected.Count > 0)
                {
                    throw new ArgumentException(
                        "Tier 3 capture does not support the following /viewport " +
                        "fields (the legacy backend honors them; future PR will " +
                        "add parity via ViewCaptureSettings): " +
                        string.Join(", ", rejected) +
                        ". Remove these fields or use captureBackend='legacy'.");
                }
            }

            return new Tier3Request
            {
                Width = ClampDimension(GetIntArg(args, "width"), DefaultWidth),
                Height = ClampDimension(GetIntArg(args, "height"), DefaultHeight),
                ViewName = GetStringArg(args, "view"),
                DisplayMode = GetStringArg(args, "displayMode"),
                ZoomExtents = GetBoolArg(args, "zoomExtents"),
                RaytracedConverge = GetBoolArg(args, "raytracedConverge"),
                RaytracedTimeoutMs = ClampRaytracedTimeoutMs(
                    GetIntArg(args, "raytracedTimeoutMs")),
            };
        }

        internal static int ClampRaytracedTimeoutMs(int? raw)
        {
            if (!raw.HasValue || raw.Value <= 0)
            {
                return DefaultRaytracedTimeoutMs;
            }
            if (raw.Value > MaxRaytracedTimeoutMs)
            {
                return MaxRaytracedTimeoutMs;
            }
            return raw.Value;
        }

        internal static int ClampDimension(int? raw, int defaultValue)
        {
            if (!raw.HasValue) return defaultValue;
            if (raw.Value < MinDimension) return MinDimension;
            if (raw.Value > MaxDimension) return MaxDimension;
            return raw.Value;
        }

        internal static bool IsRaytracedModeName(string? modeName)
        {
            if (string.IsNullOrEmpty(modeName)) return false;
            foreach (var r in RaytracedModeNames)
            {
                if (string.Equals(modeName, r, StringComparison.OrdinalIgnoreCase))
                    return true;
            }
            return false;
        }

        // ─── Rhino-dependent helpers (live-Rhino only) ──────────────────

        internal static bool IsRaytracedMode(DisplayModeDescription? mode)
        {
            if (mode == null) return false;
            return IsRaytracedModeName(mode.EnglishName)
                || IsRaytracedModeName(mode.LocalName);
        }

        internal static DisplayModeDescription? FindDisplayMode(string modeName)
        {
            if (string.IsNullOrEmpty(modeName)) return null;
            if (string.Equals(modeName, "current", StringComparison.OrdinalIgnoreCase))
                return null;

            // Standard resolver first.
            var mode = DisplayModeDescription.FindByName(modeName);
            if (mode != null) return mode;

            var allModes = DisplayModeDescription.GetDisplayModes();

            mode = allModes.FirstOrDefault(m =>
                m.EnglishName.Equals(modeName, StringComparison.OrdinalIgnoreCase));
            if (mode != null) return mode;

            mode = allModes.FirstOrDefault(m =>
                m.LocalName.Equals(modeName, StringComparison.OrdinalIgnoreCase));
            if (mode != null) return mode;

            mode = allModes.FirstOrDefault(m =>
                m.EnglishName.StartsWith(modeName, StringComparison.OrdinalIgnoreCase) ||
                m.LocalName.StartsWith(modeName, StringComparison.OrdinalIgnoreCase));
            if (mode != null) return mode;

            if (Guid.TryParse(modeName, out var guid))
            {
                mode = allModes.FirstOrDefault(m => m.Id == guid);
                if (mode != null) return mode;
            }

            return null;
        }

        private static bool TryRestoreNamedView(
            RhinoDoc doc, Rhino.Display.RhinoView view, string viewName)
        {
            var namedViews = doc.NamedViews;
            for (int i = 0; i < namedViews.Count; i++)
            {
                if (string.Equals(
                    namedViews[i].Name, viewName, StringComparison.OrdinalIgnoreCase))
                {
                    namedViews.Restore(i, view.ActiveViewport);
                    view.Redraw();
                    return true;
                }
            }
            return false;
        }

        private static bool TrySetStandardView(
            Rhino.Display.RhinoView view, string viewName)
        {
            // Map the same set as the native handler's ViewNameToCommand in
            // ViewportHandler.cpp. Use the SDK (not RunScript) to avoid
            // mutating the active view via scripted commands.
            switch (viewName.ToLowerInvariant())
            {
                case "top":
                    view.ActiveViewport.SetProjection(
                        DefinedViewportProjection.Top, null, true);
                    break;
                case "bottom":
                    view.ActiveViewport.SetProjection(
                        DefinedViewportProjection.Bottom, null, true);
                    break;
                case "front":
                    view.ActiveViewport.SetProjection(
                        DefinedViewportProjection.Front, null, true);
                    break;
                case "back":
                    view.ActiveViewport.SetProjection(
                        DefinedViewportProjection.Back, null, true);
                    break;
                case "left":
                    view.ActiveViewport.SetProjection(
                        DefinedViewportProjection.Left, null, true);
                    break;
                case "right":
                    view.ActiveViewport.SetProjection(
                        DefinedViewportProjection.Right, null, true);
                    break;
                case "perspective":
                    view.ActiveViewport.SetProjection(
                        DefinedViewportProjection.Perspective, null, true);
                    break;
                default:
                    return false;
            }
            view.Redraw();
            return true;
        }

        // Returns the final sample count observed (or 0 if the realtime
        // display mode never reported any). Port of
        // SA_Banana.Services.ViewportCapture.WaitForRaytracedRender.
        private static int WaitForRaytracedRender(
            Rhino.Display.RhinoView view, int timeoutMs)
        {
            var start = DateTime.Now;
            int lastPass = -1;
            int stableCount = 0;

            while ((DateTime.Now - start).TotalMilliseconds < timeoutMs)
            {
                var rtm = view.RealtimeDisplayMode;
                if (rtm != null)
                {
                    var passFunc = rtm.LastRenderedPass;
                    int pass = passFunc != null ? passFunc() : 0;

                    if (pass >= RaytracedMinSamples)
                    {
                        return pass;
                    }
                    if (pass == lastPass && pass > 0)
                    {
                        stableCount++;
                        if (stableCount >= RaytracedStableThreshold)
                        {
                            return pass;
                        }
                    }
                    else
                    {
                        stableCount = 0;
                    }
                    lastPass = pass;
                }
                Thread.Sleep(RaytracedCheckIntervalMs);
            }

            // Timeout: final redraw + brief settle so buffer is ready.
            view.Redraw();
            Thread.Sleep(100);
            return Math.Max(0, lastPass);
        }

        // ─── Arg helpers (mirror NativeGhBridgeRegistrar conventions) ──

        private static string? GetStringArg(
            Dictionary<string, JsonElement>? args, string name)
        {
            if (args == null) return null;
            if (!args.TryGetValue(name, out var el)) return null;
            if (el.ValueKind == JsonValueKind.String) return el.GetString();
            return null;
        }

        private static int? GetIntArg(
            Dictionary<string, JsonElement>? args, string name)
        {
            if (args == null) return null;
            if (!args.TryGetValue(name, out var el)) return null;
            if (el.ValueKind == JsonValueKind.Number
                && el.TryGetInt32(out var v))
            {
                return v;
            }
            return null;
        }

        private static bool GetBoolArg(
            Dictionary<string, JsonElement>? args, string name)
        {
            if (args == null) return false;
            if (!args.TryGetValue(name, out var el)) return false;
            if (el.ValueKind == JsonValueKind.True) return true;
            if (el.ValueKind == JsonValueKind.False) return false;
            return false;
        }
    }
}
